File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/Community/CommunityController.php
<?php

namespace App\Http\Controllers\Api\V1\Community;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\Community\Requests\StoreCommentRequest;
use App\Modules\Community\Requests\StorePostRequest;
use App\Modules\Community\Requests\UpdatePostRequest;
use App\Modules\Community\Services\CommunityService;

class CommunityController extends ApiController
{
    public function __construct(
        private readonly CommunityService $communityService,
    ) {
    }

    public function boards()
    {
        return $this->success($this->communityService->listBoards());
    }

    public function posts(int $boardId)
    {
        return $this->success($this->communityService->listPosts($boardId));
    }

    public function showPost(int $id)
    {
        return $this->success($this->communityService->showPost($id));
    }

    public function storePost(StorePostRequest $request, int $boardId)
    {
        return $this->success($this->communityService->createPost($request->user(), $boardId, $request->validated()), status: 201);
    }

    public function updatePost(UpdatePostRequest $request, int $id)
    {
        return $this->success($this->communityService->updatePost($request->user(), $id, $request->validated()));
    }

    public function destroyPost(int $id)
    {
        return $this->success($this->communityService->deletePost(request()->user(), $id));
    }

    public function toggleLike(int $id)
    {
        return $this->success($this->communityService->toggleLike(request()->user(), $id));
    }

    public function comments(int $id)
    {
        return $this->success($this->communityService->listComments($id));
    }

    public function storeComment(StoreCommentRequest $request, int $id)
    {
        return $this->success($this->communityService->createComment($request->user(), $id, $request->validated()), status: 201);
    }

    public function destroyComment(int $id)
    {
        return $this->success($this->communityService->deleteComment(request()->user(), $id));
    }
}

File: BackEnd/Java/springshop_maven_jpa/src/main/java/com/pbshop/springshop/community/CommunityController.java
package com.pbshop.springshop.community;

import java.util.List;
import java.util.Map;

import jakarta.validation.Valid;

import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.pbshop.springshop.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.springshop.common.api.ApiResponse;
import com.pbshop.springshop.community.dto.CommunityDtos.SaveCommentRequest;
import com.pbshop.springshop.community.dto.CommunityDtos.SavePostRequest;

@RestController
@RequestMapping("${pbshop.api.base-path:/api/v1}")
public class CommunityController {

    private final CommunityService communityService;

    public CommunityController(CommunityService communityService) {
        this.communityService = communityService;
    }

    @GetMapping("/boards")
    public ApiResponse<List<Map<String, Object>>> getBoards() {
        return ApiResponse.success(communityService.getBoards());
    }

    @GetMapping("/boards/{boardId}/posts")
    public ApiResponse<Map<String, Object>> getBoardPosts(
            @PathVariable Long boardId,
            @RequestParam(required = false) String search,
            @RequestParam(defaultValue = "newest") String sort,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit
    ) {
        return ApiResponse.success(communityService.getBoardPosts(boardId, search, sort, page, limit));
    }

    @GetMapping("/posts/{id}")
    public ApiResponse<Map<String, Object>> getPost(@PathVariable Long id) {
        return ApiResponse.success(communityService.getPost(id));
    }

    @PostMapping("/boards/{boardId}/posts")
    public ApiResponse<Map<String, Object>> createPost(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long boardId,
            @Valid @RequestBody SavePostRequest request
    ) {
        return ApiResponse.success(communityService.createPost(principal, boardId, request));
    }

    @PatchMapping("/posts/{id}")
    public ApiResponse<Map<String, Object>> updatePost(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id,
            @Valid @RequestBody SavePostRequest request
    ) {
        return ApiResponse.success(communityService.updatePost(principal, id, request));
    }

    @DeleteMapping("/posts/{id}")
    public ApiResponse<Map<String, Object>> deletePost(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id
    ) {
        return ApiResponse.success(communityService.deletePost(principal, id));
    }

    @PostMapping("/posts/{id}/like")
    public ApiResponse<Map<String, Object>> toggleLike(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id
    ) {
        return ApiResponse.success(communityService.toggleLike(principal, id));
    }

    @GetMapping("/posts/{id}/comments")
    public ApiResponse<List<Map<String, Object>>> getComments(@PathVariable Long id) {
        return ApiResponse.success(communityService.getComments(id));
    }

    @PostMapping("/posts/{id}/comments")
    public ApiResponse<Map<String, Object>> createComment(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id,
            @Valid @RequestBody SaveCommentRequest request
    ) {
        return ApiResponse.success(communityService.createComment(principal, id, request));
    }

    @DeleteMapping("/comments/{id}")
    public ApiResponse<Map<String, Object>> deleteComment(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id
    ) {
        return ApiResponse.success(communityService.deleteComment(principal, id));
    }
}
